### PR TITLE
[#259] Add implementation of REST interface for device registration.

### PR DESCRIPTION
**Pull request by b-abel**
The implementation contains only the GET operation and the interface
is not yet exposed.

Signed-off-by: Abel Buechner <Abel.Buechner@bosch-si.com>

### CODE REVIEW

File: service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Base implementation for a REST interface which can be used to register devices.
+ *
+ * TODO: add all operations of the Registration API.
+ */
+public class DeviceRegistryRestServerBase extends HttpServiceBase<ServiceConfigProperties> {
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: I think we should keep this open for using config classes extending `ServiceConfigProperties`.

Comment (b-abel) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: You think of something like the following?
`public class DeviceRegistryRestServerBase<T> extends HttpServiceBase<T extends ServiceConfigProperties> { 
...
}`

Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: yes, we are using this kind of notation on other classes as well, however, we I think it needs to be the other way around:

`public class DeviceRegistryRestServerBase<T extends ServiceConfigProperties> extends HttpServiceBase<T> {...}`

Comment (b-abel) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: Right. Done.

File: service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Autowired
+    @Qualifier(CONFIGURATION_QUALIFIER)
+    @Override
+    public void setConfig(final ServiceConfigProperties configuration) {
+        setSpecificConfig(configuration);
+    }
+
+    @Override
+    protected void addRoutes(Router router) {
```
Comment (sysexcontrol) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: final Router router... ?

Comment (b-abel) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: Done.

File: service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Base implementation for a REST interface which can be used to register devices.
+ *
+ * TODO: add all operations of the Registration API.
+ */
+public class DeviceRegistryRestServerBase<T extends ServiceConfigProperties> extends HttpServiceBase<T> {
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: the JavaDoc should also contain a `@param` line for T

File: service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(DeviceRegistryRestServerBase.class);
+
+    private static final String PARAM_TENANT = "tenant";
+    private static final String PARAM_DEVICE_ID = "device_id";
+
+    @Autowired
+    @Qualifier(CONFIGURATION_QUALIFIER)
+    @Override
+    public void setConfig(final T configuration) {
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: do we want anybody to override this method? If so, why?

File: service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        setSpecificConfig(configuration);
+    }
+
+    @Override
+    protected void addRoutes(final Router router) {
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: do we want anybody to override this method? If so, why?

File: service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java (hunk 4 above)
Comment (b-abel) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: I don't see a use case for that. So I'll make it final.

File: service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java (hunk 5 above)
Comment (b-abel) [inline service-base/src/main/java/org/eclipse/hono/service/registration/DeviceRegistryRestServerBase.java]: I don't see a use case for that. So I'll make it final.

**Comment by sophokles73** [comment]
@b-abel can you squash and rebase?

**Comment by b-abel** [comment]
Rebased and squashed.

**Pull request merged**